core/embed/trezorhal/stm32f4/syscall_verifiers.c

```diff
@@ -0,0 +1,667 @@
+                                  size_t footer_len) {
+  char title_copy[64] = {0};
+  char message_copy[64] = {0};
+  char footer_copy[64] = {0};
+
+  title_len = title ? MIN(title_len, sizeof(title_copy) - 1) : 0;
+  message_len = message ? MIN(message_len, sizeof(message_copy) - 1) : 0;
+  footer_len = footer ? MIN(footer_len, sizeof(footer_copy) - 1) : 0;
+
+  if (title != NULL) {
+    if (!probe_read_access(title, title_len)) {
+      goto access_violation;
+    }
+
+    title = strcpy(title_copy, title);
+  }
+
+  if (message != NULL) {
+    if (!probe_read_access(message, message_len)) {
+      goto access_violation;
+    }
+
+    message = strcpy(message_copy, message);
+  }
+
+  if (footer != NULL) {
+    if (!probe_read_access(footer, footer_len)) {
+      goto access_violation;
+    }
+
+    footer = strcpy(footer_copy, footer);
+  }
+
+  systask_exit_error(task, title, title_len, message, message_len, footer,
+                     footer_len);
+
+  return;
+
+access_violation:
+  apptask_access_violation();
+}
+
+void systask_exit_fatal__verified(systask_t *task, const char *message,
+                                  size_t message_len, const char *file,
+                                  size_t file_len, int line) {
+  char message_copy[64] = {0};
+  char file_copy[64] = {0};
+
+  message_len = message ? MIN(message_len, sizeof(message_copy) - 1) : 0;
+  file_len = file ? MIN(file_len, sizeof(file_copy) - 1) : 0;
+
+  if (message != NULL) {
+    if (!probe_read_access(message, message_len)) {
+      goto access_violation;
+    }
+
+    message = strcpy(message_copy, message);
+  }
+
+  if (file != NULL) {
+    if (!probe_read_access(file, file_len)) {
+      goto access_violation;
+    }
+
+    file = strcpy(file_copy, file);
+  }
+
+  systask_exit_fatal(task, message, message_len, file, file_len, line);
+
+  return;
+
+access_violation:
+  apptask_access_violation();
+}
+
+// ---------------------------------------------------------------------
+
+void reboot_and_upgrade__verified(const uint8_t hash[32]) {
+  if (!probe_read_access(hash, 32)) {
+    goto access_violation;
+  }
+
+  reboot_and_upgrade(hash);
+
+access_violation:
+  apptask_access_violation();
+}
+
+// ---------------------------------------------------------------------
+
+#ifdef XFRAMEBUFFER
+
+bool display_get_frame_buffer__verified(display_fb_info_t *fb) {
+  if (!probe_write_access(fb, sizeof(*fb))) {
+    goto access_violation;
+  }
+
+  display_fb_info_t fb_copy = {0};
+
+  bool result = display_get_frame_buffer(&fb_copy);
+
+  *fb = fb_copy;
+
+  return result;
+
+access_violation:
+  apptask_access_violation();
+  return false;
+}
+
+#endif  // XFRAMEBUFFER
+
+void display_fill__verified(const gfx_bitblt_t *bb) {
+  if (!probe_read_access(bb, sizeof(*bb))) {
+    goto access_violation;
+  }
+
+  display_fill(bb);
+
+  return;
+
+access_violation:
+  apptask_access_violation();
+}
+
+void display_copy_rgb565__verified(const gfx_bitblt_t *bb) {
```

Comment:
the bitblt structure should be copied and only after that verified, as it can be modified later be the applet. (for example by dma2d operation setup before calling this function, or maybe some callbacks)

(in the fill funciton too)

Reply:
You are right, see this fix b2ee71c35db10524c5d78c20d4777c3d3b99b982